### ISSUE TITLE
[data] IterableDataset drops samples with multiple DataLoader workers

### ISSUE THREAD

**Issue by yeyeyeping** [ByteDance-Seed/VeOmni#1240]
## Problem

When VeOmni's `ShardedIterableDataset` wraps a Hugging Face `IterableDataset`, using multiple DataLoader workers can silently drop records.

For example, with 48 records, one source shard, and DP size 1, setting `num_workers=2` returns only 24 records. All 48 records should be returned exactly once.

## Cause

An HF `IterableDataset` already partitions its source shards when iterated inside a PyTorch DataLoader worker. This is implemented by HF itself; it is not automatic behavior for every PyTorch `IterableDataset`.

In datasets 2.21, [`IterableDataset.__iter__`](https://github.com/huggingface/datasets/blob/2.21.0/src/datasets/iterable_dataset.py#L2004-L2012) checks whether it is running inside a worker and dispatches to `_iter_pytorch()`:

```python
worker_info = torch.utils.data.get_worker_info()
if isinstance(self, torch.utils.data.IterableDataset) and worker_info is not None:
    yield from self._iter_pytorch()
```

Then [`_iter_pytorch()`](https://github.com/huggingface/datasets/blob/2.21.0/src/datasets/iterable_dataset.py#L1907-L1914) assigns source shards to that worker:

```python
ex_iterable = ex_iterable.shard_data_sources(worker_id=worker_info.id, num_workers=worker_info.num_workers)
```

However, VeOmni applies another worker filter to this already partitioned stream:

```python
index % self._dp_size == self._dp_rank and (index // self._dp_size) % num_workers == worker_id
```

The second filter drops records that no other worker will return. With one shard and two workers, HF assigns the shard to worker 0 and stops worker 1. VeOmni then keeps only every other record from worker 0.

## Minimal reproduction

Verified on CPU with Python 3.12.0, `datasets==2.21.0`, and `torch==2.13.0+cu130`. No GPU or distributed process group is required.

```python
from datasets import Dataset
from torch.utils.data import DataLoader

from veomni.data.dataset import ShardedIterableDataset


def main():
    for workers in (0, 2):
        raw = Dataset.from_dict({"id": list(range(48))})
        raw = raw.to_iterable_dataset(num_shards=1)
        dataset = ShardedIterableDataset(
            raw, dp_rank=0, dp_size=1, repeat=False
        )
        rows = list(DataLoader(dataset, batch_size=None, num_workers=workers))
        print(workers, len(rows), [row["id"] for row in rows])


if __name__ == "__main__":
    main()
```

Expected: 48 records in both cases.

Observed:

- `num_workers=0`: 48 records, IDs `0..47`.
- `num_workers=2`: 24 records, IDs `0, 2, 4, ..., 46`.

DP size is 1 and batching is disabled, so neither DP tail dropping nor batch `drop_last` explains the missing records.

### 1. Remove the redundant worker filter for HF datasets

Let HF handle worker partitioning. Keep VeOmni's row-level DP filtering, incomplete-DP-round dropping, and repeat loop. Retain worker partitioning for non-HF iterables where needed.

This is the smallest fix. It retains the current I/O cost of every DP rank scanning the source data. It also retains the limitation that a worker stream shorter than `dp_size` yields no complete DP round, even with repeat enabled.

### 2. Use HF concatenation and sharding with the current dependency version

Replace the custom sharding/repeat wrapper with:

```python
ds = concatenate_datasets([raw_dataset] * repeat_count)
ds = split_dataset_by_node(ds, rank=dp_rank, world_size=dp_size)
```

HF handles both DP and worker partitioning. This provides finite repetition without changing the current `datasets>=2.20.0,<=2.21.0` constraint.

On datasets 2.21, concatenate **before** node splitting; concatenating already node-split datasets loses their distributed configuration. The repetition count must be chosen to supply enough batches on every rank. Finite repetition does not equalize unequal partitions, so this is not a direct replacement for unlimited `dataset_repeat=True`.

### 3. Upgrade datasets and use native repeat with HF sharding

Use a version supporting [`IterableDataset.repeat`](https://huggingface.co/docs/datasets/en/package_reference/main_classes#datasets.IterableDataset.repeat), replacing the custom wrapper with:

```python
ds = raw_dataset.repeat(None)  # Or repeat(N) for finite repetition.
ds = split_dataset_by_node(ds, rank=dp_rank, world_size=dp_size)
```

With unlimited repetition, streams that keep producing valid batches can supply each rank until the trainer reaches `train_steps`. Smaller partitions repeat more frequently; empty or fully filtered streams still need handling. The upgrade would also require compatibility, shuffle, and checkpoint/resume validation.

For **options 2 and 3**, HF [assigns whole shards when the shard count is divisible by DP size](https://huggingface.co/docs/datasets/v2.21.0/en/package_reference/main_classes#datasets.distributed.split_dataset_by_node), preserving shard-level I/O efficiency. Otherwise, including when shards are fewer than ranks, it falls back to example-level skipping. Workers still receive source shards; excess workers stop. Repetition does not increase the source shard count.

These options do not automatically preserve the current per-pass DP tail-dropping rule. Repeated streams can cross pass boundaries during example-level splitting. With finite repetition or `dataset_repeat=False`, unequal rank batch counts still need an exhaustion policy. Repeat does not change VeOmni's calculated per-epoch step budget.

Which approach would be preferred? Which guarantees should we preserve for finite iteration and repeat mode, especially per-pass equalization, tail dropping, and reaching the configured step budget?

**Comment by Coach257** [comment]
Thanks for the detailed report — reproduced on current main (`datasets==2.21.0`). With `num_workers=N`, each DP rank receives only `1/N` of the rows, regardless of shard count or DP size (48 rows → 24 with 2 workers, 12 with 4). `dataloader.num_workers` defaults to 2, so the default `iterable` / `interleave` paths are affected. It is a regression from #1156, which replaced `split_dataset_by_node` with `ShardedIterableDataset`. The existing tests only use `_ListStream` with `num_workers=0`, so they don't catch it.

## Preferred direction: Option 3 (upgrade `datasets`, let HF do DP/worker sharding + resume)

Target semantics:
- **Iterable datasets are always infinite; an epoch ends at `train_steps`.** This is how we guarantee all DP ranks stop at the same step without trainer changes: the trainer catches `StopIteration` per rank and breaks locally (`veomni/trainer/base.py:819-824`), so any per-rank count mismatch in a finite stream hangs the other ranks in FSDP collectives. `dataset_repeat=False` would no longer mean "exactly one pass"; it needs a decision whether to remove or redefine it.
- **Each repeat pass must be reshuffled.**
- **Checkpoint resume must work across pass boundaries.**
- Shard balance is the dataset producer's responsibility. Idle workers when `num_shards < num_workers` are acceptable.

## Findings from prototyping with `datasets==5.0.1` (please account for these)

1. **Native `repeat()` does not reshuffle between passes.** `RepeatExamplesIterable.__iter__` does `yield from self.ex_iterable` on the same iterable, and shard order is shuffled once at iteration start.
   - `shuffle(...).repeat(None)`: every pass has an identical order. Only an outer `set_epoch` changes it.
   - `repeat(None).shuffle(...)`: the buffer mixes across pass boundaries, but the shard order is still fixed across passes.

   So a thin per-pass loop is still needed: for each pass, `set_epoch(seed + epoch + pass_idx)`, then iterate.

2. **`set_epoch` must not be called on the shared dataset inside workers.** Both in 2.21 and 5.x, `IterableDataset._epoch` is a shared-memory tensor shared by all DataLoader workers, and `set_epoch` updates it in place. Calling it per pass inside each worker lets workers overwrite each other's epoch and derive different shard orders, so one shard gets read twice and another skipped. On 2.21 (2 files, 2 workers, 16 passes/worker): 10/160 passes corrupted. With a `copy.deepcopy` of the dataset per worker at the start of `__iter__`: 0/160. This is also a live bug on main today (`ShardedIterableDataset.__iter__` calls `self._dataset.set_epoch(...)` per pass).

3. **~5 s stall per pass with parquet + shuffle on 5.x.** The parquet builder sets `SLEEP_ON_THREADS_SHUTDOWNS = True` (a workaround for apache/arrow#45214), and the multi-source iterator sleeps `config.SLEEP_TIME_ON_THREADS_SHUTDOWN = 5` s every time a pass ends. Measured: 8-row dataset, `shuffle=True` → cumulative `[5, 10, 15, 20]` s after passes 1-4; `shuffle=False` → ~0 s. This is negligible for large dumps but blocks the small-dump case `dataset_repeat` exists for (e.g. MiniMax H3 offline, 8 rows on dp=8 → ~5 s per step). 2.21 does not have this sleep. Please check whether it can be avoided safely (setting the config to 0 disables the Arrow workaround; risk unknown).

4. **Finite streams are not equal across ranks with HF sharding**, even with balanced data. Examples: `[61, 60, 60]`, and `[14, 14, 13]` on 2 files / dp=3 / 2 workers. Row-level fallback applies the step inside each worker's sub-stream, so ranks can differ by up to `num_workers` rows. `repeat(None).take(n)` does not fix this: `take` is applied per worker (each rank got `2n` with 2 workers), and on the row-level path it is applied before the DP step (`[10, 8, 8]`). Hence the "always infinite, stop at `train_steps`" semantics above.

5. **Resume.** HF `state_dict` is not exact with shuffle: the buffer contents are not saved (30–53 of 60 samples mismatched after resume, dp=2, workers 0/2). Exact resume works with a replay-skip state instead. Each worker saves `{pass_idx, n_yielded_in_pass}`; on resume it calls `set_epoch(seed + epoch + pass_idx)` on its private copy, re-iterates the pass, and skips `n` rows. This works because HF shuffling is deterministic for a given seed/epoch. Measured 0/60 mismatches with shuffle on, workers 0/2, resuming both mid-first-pass and mid-second-pass. Cost: up to one pass of re-reading per worker on resume (before transforms). Optionally, use HF `state_dict` when shuffle is off, since it is exact and O(1) in that case. Main today has no iterable resume at all (`ShardedIterableDataset` lacks `state_dict`, so training restarts from row 0).

6. **Upgrade cost.** `repeat()` exists since 3.3 but breaks with `split_dataset_by_node` until 4.0 (huggingface/datasets#7561), so the floor is `>=4.0`. Installing 5.0.1 pulled `packaging==26.3`, which conflicts with our `packaging<26.0`. Please go through the uv update workflow and run cross-modality regression. VeOmni only uses `load_dataset` / `interleave_datasets` / `IterableDataset` directly, so the code surface is small.

7. **Worth evaluating before committing to the upgrade:** because of (1) and (2), a per-pass wrapper is needed anyway. `split_dataset_by_node` + that wrapper + HF `state_dict` also works on 2.21, which avoids (3) and (6). Native `repeat()` then buys little. Please compare both and state the choice in the PR.

## Suggested PR scope
- Delete `ShardedIterableDataset`. DP + worker sharding → `split_dataset_by_node`; per-pass DP tail drop is no longer needed (infinite stream, bounded by `train_steps`). Fold the remaining per-pass loop (per-worker `deepcopy`, `set_epoch(seed + epoch + pass_idx)` each pass, `{pass_idx, n}` replay-skip `state_dict` / `load_state_dict`) into `IterativeDataset`, so `InterleavedIterableDataset` inherits it. A worker whose pass yields nothing just returns, since excess workers are stopped by HF by design. Fail at build time instead if total rows < `dp_size`.
- Tests with a real HF streaming parquet dataset, `StatefulDataLoader`, and `num_workers ∈ {0, 2}`:
  - no loss/duplication within a pass;
  - different order in each pass;
  - no cross-worker shard overlap;
  - exact resume mid-pass and across a pass boundary (shuffle off);
  - all ranks keep producing until `train_steps`.
- Docs: iterable is infinite and bounded by `train_steps`; shards should be balanced; excess workers idle.

Out of scope: equal batch counts under `dyn_bsz` or filtering transforms; `persistent_workers=True` not propagating the wrapper's epoch.
